File: src/infinicore/ops/qwen3_rms_norm/qwen3_rms_norm.cc
#include "infinicore/ops/qwen3_rms_norm.hpp"

#include "infinicore/ops/rms_norm.hpp"

namespace infinicore::op {

Tensor qwen3_rms_norm(const Tensor &x, const Tensor &weight, float epsilon) {
    return rms_norm(x, weight, epsilon);
}

void qwen3_rms_norm_(Tensor y, const Tensor &x, const Tensor &weight, float epsilon) {
    rms_norm_(y, x, weight, epsilon);
}

} // namespace infinicore::op

File: python/infinicore/ops/qwen3_rms_norm.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def qwen3_rms_norm(x: Tensor, weight: Tensor, epsilon: float = 1e-6, *, out: Tensor | None = None) -> Tensor:
    if out is None:
        return Tensor(_infinicore.qwen3_rms_norm(x._underlying, weight._underlying, epsilon))
    _infinicore.qwen3_rms_norm_(out._underlying, x._underlying, weight._underlying, epsilon)
    return out
